### PR TITLE
Implement row alias expressions (`INSERT ... VALUES (...) AS new_tbl ON DUPLICATE x = new_tbl.x`)

### PR DESCRIPTION
**Pull request by nicktobey**
When inserting values, the user can specify names for both the source table and columns which are used in `ON DUPLICATE` expressions. It looks like either of the below options:

```sql
INSERT INTO tbl VALUES (1, 2) AS tbl_new ON DUPLICATE KEY b = tbl_new.b;

INSERT INTO tbl VALUES (1, 2) AS tbl_new(a_new, b_new) ON DUPLICATE KEY b = b_new;
```

This replaces the previous (now-deprecated) syntax:

```sql
INSERT INTO tbl VALUES (1, 2) ON DUPLICATE KEY b = VALUES(b);
```

Supporting both syntaxes together was non-trivial because it means there's now two different ways to refer to the same column. While he had an existing way to "redirect" one column name to another, this only worked for unqualified names (no table name), and it overrode the normal name resolution rules, which meant we would fail to detect cases that should be seen as ambiguous.

Previously, we would implement references to the inserted values by using a special table named "__new_ins". I implemented this by keeping that as the default, but using the row alias instead of one was provided. We then create a map from the destination table names to column aliases, and use that map to rewrite expressions that appear inside the VALUES() function.

### CODE REVIEW

**Comment by nicktobey** [comment]
Fixes https://github.com/dolthub/dolt/issues/7638

File: sql/planbuilder/builder.go [hunk 1]
```diff
@@ -34,21 +34,23 @@ var BinderFactory = &sync.Pool{New: func() interface{} {
...
+	colId               columnId
+	tabId               sql.TableId
+	multiDDL            bool
+	viewCtx             *ViewContext
+	procCtx             *ProcContext
+	triggerCtx          *TriggerContext
+	bindCtx             *BindvarContext
+	insertActive        bool
+	insertTableAlias    string
+	insertColumnAliases map[string]string
```
Comment (max-hoffman) [inline sql/planbuilder/builder.go:51]: can we limit these variables to a single scope? the rest of these objects explicitly outlive a single query

File: sql/planbuilder/scalar.go [hunk 2]
```diff
@@ -316,7 +316,10 @@ func (b *Builder) buildScalar(inScope *scope, e ast.Expr) (ex sql.Expression) {
 	case *ast.ValuesFuncExpr:
 		if b.insertActive {
 			if v.Name.Qualifier.Name.String() == "" {
-				v.Name.Qualifier.Name = ast.NewTableIdent(OnDupValuesPrefix)
+				v.Name.Qualifier.Name = ast.NewTableIdent(b.insertTableAlias)
+				if len(b.insertColumnAliases) > 0 {
+					v.Name.Name = ast.NewColIdent(b.insertColumnAliases[v.Name.Name.String()])
```
Comment (max-hoffman) [inline sql/planbuilder/scalar.go:321]: strings to lower?

File: sql/planbuilder/dml.go [hunk 3]
```diff
@@ -97,8 +117,13 @@ func (b *Builder) buildInsert(inScope *scope, i *ast.Insert) (outScope *scope) {
 			if len(srcScope.cols) == len(destScope.cols) {
 				combinedScope.newColumn(srcScope.cols[i])
 			} else {
-				// check for VALUES refs
-				c.table = OnDupValuesPrefix
+				// The to-be-inserted values can be referenced via the provided alias.
+				c.table = b.insertTableAlias
+				if len(b.insertColumnAliases) > 0 {
+					aliasColumnName := b.insertColumnAliases[c.col]
```
Comment (max-hoffman) [inline sql/planbuilder/dml.go:123]: strings to lower?

File: sql/planbuilder/dml.go [hunk 4]
```diff
@@ -85,7 +84,28 @@ func (b *Builder) buildInsert(inScope *scope, i *ast.Insert) (outScope *scope) {
...
+	if aliasedValues, ok := insertRows.(*ast.AliasedValues); ok {
+		valueTableName := aliasedValues.As.String()
+		if valueTableName != "" {
+			b.insertTableAlias = valueTableName
+		}
+		if len(aliasedValues.Columns) > 0 {
+			b.insertColumnAliases = make(map[string]string)
+			for i, destColumn := range columns {
+				sourceColumn := aliasedValues.Columns[i].String()
+				b.insertColumnAliases[destColumn] = sourceColumn
```
Comment (max-hoffman) [inline sql/planbuilder/dml.go:103]: lower strings?

**Review by max-hoffman — APPROVED** [review]
LGTM, I don't love making ON DUP EXPRS more complicated, but there aren't a lot of faster practical alternatives. Like most of the complexity around INSERTs, we should probably do something that reorders/normalizes insert value ordering to avoid extra complexity in the rest of analysis.

File: sql/planbuilder/builder.go (hunk 1 above)
Comment (nicktobey) [reply sql/planbuilder/builder.go]: Limited it to two scopes (insert scope and combined scope)

File: sql/planbuilder/dml.go (hunk 3 above)
Comment (nicktobey) [reply sql/planbuilder/dml.go]: Lowered the string on insert into the map.

File: sql/planbuilder/dml.go (hunk 4 above)
Comment (nicktobey) [reply sql/planbuilder/dml.go]: Done.

**Pull request merged**